Spam learn

Hi,
is this a bug?

`
2021-06-11T10:09:27.741: [ERROR   ] spamd - Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kopano/log.py", line 138, in log_exc
    try: yield
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 81, in update
    self.learn(item, searchkey, True)
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 103, in learn
    gid = grp.getgrnam(self.sagroup).gr_gid
KeyError: "getgrnam(): name not found: 'amavis'"`

the database is not growing

Thanks

Hello fbartels, do you mean the following. 

**Describe the bug**
If mails are pushed into the Junk folder, they are not learned. The following error occurs in the Docker log:

Notice: Container is run read-only, skipping package installation.
If you want to have additional packages installed in the container either:
- Build your own image with the packages already included
- switch the container to 'read_only: false
Jun 11 09:33:32 Configure core service 'spamd'
Using Kopano Groupware Core: 11.0.0.1.2bb6991ac-0 + 432.1
2021/06/11 09:33:32 Ready: file: /// var / lib / dbus / machine-id.
2021/06/11 09:33:32 Ready: file: /// etc / machine-id.
2021/06/11 09:33:32 Ready: file: ///var/run/kopano/server.sock.
2021-06-11T10: 09: 27.741: [ERROR] spamd - Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kopano/log.py", line 138, in log_exc
    try: yield
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 81, in update
    self.learn (item, searchkey, True)
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 103, in learn
    gid = grp.getgrnam (self.sagroup) .gr_gid
KeyError: "getgrnam (): name not found: 'amavis'" 


**To Reproduce**

when running manually in the Docker container kopano_spamd, the error occurs: 

root@2c2b533dfbab:/kopano/path# kopano-spamd -l DEBUG
2021-06-11T11:02:12.808: [INFO    ] spamd - starting spamd
2021-06-11T11:08:15.784: [INFO    ] spamd - Learning message as SPAM, entryid: 0000000027237074396243CB89AC4950448269BE0100000005000000770AB7524E6D48538BBEB99D83BC36EF00000000
2021-06-11T11:08:15.795: [ERROR   ] spamd - Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kopano/log.py", line 138, in log_exc
    try: yield
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 81, in update
    self.learn(item, searchkey, True)
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 103, in learn
    gid = grp.getgrnam(self.sagroup).gr_gid
KeyError: "getgrnam(): name not found: 'amavis'"

2021-06-11T11:08:15.826: [INFO    ] spamd - Learning message as SPAM, entryid: 0000000027237074396243CB89AC4950448269BE01000000050000009057E2BB77B140098D047500F36927AB00000000
2021-06-11T11:08:15.838: [ERROR   ] spamd - Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kopano/log.py", line 138, in log_exc
    try: yield
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 81, in update
    self.learn(item, searchkey, True)
  File "/usr/lib/python3/dist-packages/kopano_spamd/__init__.py", line 103, in learn
    gid = grp.getgrnam(self.sagroup).gr_gid
KeyError: "getgrnam(): name not found: 'amavis'"

**Expected behavior**
learn the spam and delete it if necessary

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -516,29 +516,6 @@ services:
     tmpfs:
       - /tmp
 
-  kopano_spamd:
-    image: ${docker_repo:-zokradonh}/kopano_core:${CORE_VERSION:-latest}
-    read_only: true
-    restart: unless-stopped
-    container_name: ${COMPOSE_PROJECT_NAME}_spamd
-    depends_on:
-      - kopano_server
-    volumes:
-      - /etc/machine-id:/etc/machine-id
-      - /etc/machine-id:/var/lib/dbus/machine-id
-      - kopanosocket/:/run/kopano
-      - kopanossl/:/kopano/ssl
-      - kopanospamd/:/var/lib/kopano/spamd
-    environment:
-      - SERVICE_TO_START=spamd
-      - TZ=${TZ}
-    env_file:
-      - kopano_spamd.env
-    networks:
-      - kopano-net
-    tmpfs:
-      - /tmp
-
 
 volumes:
   kdavstates:
@@ -546,7 +523,6 @@ volumes:
   kopanograpi:
   kopanolicenses:
   kopanosocket:
-  kopanospamd:
   kopanossl:
   kopanowebapp:
   web:

--- a/setup.sh
+++ b/setup.sh
@@ -41,7 +41,7 @@ if [ ! -e /etc/machine-id ]; then
 fi
 
 echo "Creating individual env files for containers (if they do not exist already)"
-for dockerenv in ldap password-self-service mail db kopano_ssl kopano_server kopano_webapp kopano_zpush kopano_grapi kopano_kapi kopano_dagent kopano_spooler kopano_gateway kopano_ical kopano_monitor kopano_scheduler kopano_search kopano_konnect kopano_kwmbridge kopano_kwmserver kopano_meet kopano_kapps kopano_spamd; do
+for dockerenv in ldap password-self-service mail db kopano_ssl kopano_server kopano_webapp kopano_zpush kopano_grapi kopano_kapi kopano_dagent kopano_spooler kopano_gateway kopano_ical kopano_monitor kopano_scheduler kopano_search kopano_konnect kopano_kwmbridge kopano_kwmserver kopano_meet kopano_kapps; do
 	touch ./"$dockerenv".env
 done
 

--- a/spamd-extras/README.md
+++ b/spamd-extras/README.md
@@ -1,0 +1,31 @@
+# kopano-spamd extras for kopano-docker
+
+This directory contains a compose file including optional containers to enable learning spam to be used with the [SpamAssassin](https://spamassassin.apache.org/) bayes filter.
+
+## How to use this compose file?
+
+1. Add the `spamd-extras.yml` to the `COMPOSE_FILE` variable in your `.env` file.
+
+Example:
+
+```bash
+COMPOSE_FILE=docker-compose.yml:docker-compose.ports.yml:spamd-extras/spamd-extras.yml
+```
+
+2. Run `docker-compose up -d`.
+
+
+## kopano-spamd
+
+After startup there will be a new service `kopano-spamd` which will persist mails that are moved to `Junk` to a folder named `spam` inside the `kopanospamd` volume.
+Likewise mails that are moved from the `Junk` back to `Inbox` are persisted in `ham`. Both folders indicate mails that should be trained as either being ham or spam.
+
+The `kopano-scheduler` container is extended to run the training inside the `mail` docker container at about 4am with training results being picked up by SpamAssassin
+automatically. You can check the docker logs of the scheduler for the results of each run.
+
+While already trained files can be deleted immediately after each training run, there is no cleanup provided here. The `kopanospamd` volume will therefore grow over time.
+
+For the bayes filter to start working you will need to train at least 200 mails of each ham and spam. To create a set of initial data you can use the Kopano WebApp
+by selecting mails and using the `Export as` function to create a zip file of those mails and put them into the appropriate folder.
+
+Read more about how to create effective training data here: https://spamassassin.apache.org/full/3.4.x/doc/sa-learn.html#EFFECTIVE-TRAINING

--- a/spamd-extras/spamd-extras.yml
+++ b/spamd-extras/spamd-extras.yml
@@ -1,0 +1,37 @@
+version: "3.5"
+
+services:
+  kopano_spamd:
+    image: ${docker_repo:-zokradonh}/kopano_core:${CORE_VERSION:-latest}
+    read_only: true
+    restart: unless-stopped
+    container_name: ${COMPOSE_PROJECT_NAME}_spamd
+    depends_on:
+      - kopano_server
+      - mail
+    volumes:
+      - /etc/machine-id:/etc/machine-id
+      - /etc/machine-id:/var/lib/dbus/machine-id
+      - kopanosocket/:/run/kopano
+      - kopanossl/:/kopano/ssl
+      - kopanospamd/:/var/lib/kopano/spamd
+    environment:
+      - SERVICE_TO_START=spamd
+      - TZ=${TZ}
+      - KCCONF_SPAMD_SA_GROUP=kopano
+    networks:
+      - kopano-net
+    tmpfs:
+      - /tmp
+
+  mail:
+    volumes:
+      - kopanospamd/:/var/lib/kopano/spamd
+
+  kopano_scheduler:
+    environment:
+      - CRONDELAYED_LEARN_HAM=0 4 * * * docker exec kopano_mail sa-learn --spam /var/lib/kopano/spamd/spam --dbpath /var/mail-state/lib-amavis/.spamassassin
+      - CRONDELAYED_LEARN_SPAM=15 4 * * * docker exec kopano_mail sa-learn --ham /var/lib/kopano/spamd/ham --dbpath /var/mail-state/lib-amavis/.spamassassin
+
+volumes:
+  kopanospamd:

